blog/2019-03-28-debezium-at-pipedrive.adoc

```diff
@@ -0,0 +1,225 @@
+** If it doesn’t, create it using HTTP POST request
+** If it does, but existing configuration doesn’t match our rendered one, update existing using PUT request
+** do nothing if connector exists and has latest configuration.
+
+We added these new environment variables to Debezium official container.
+
+*JOB_AUTO_CREATE* - on startup renders job template and creates or updates connector using localhost REST API.
+
+*JOB_AUTO_RESTART* - try to restart failed connector on failure
+
+*JOB_TEMPLATE* - which json file to take load as connector configuration template.
+
++ other env variables that can be referenced from the template.
+
+The job templates are stored inside container image on build time. Templates allowed us to build version of the Debezium image once and reuse it on various MySQL server setups. Templating also has conditionals support to enable some parameters only in specific environment. No more manual curl’ing.
+
+Example connector job template:
+
+[source,json]
+----
+{
+ "tasks.max": "1",
+ "connector.class": "io.Debezium.connector.MySQL.MySQLConnector",
+
+ "database.hostname": "{MySQL.host}",
+ "database.port": "{MySQL.port}",
+ "database.user": "{MySQL.username}",
+ "database.password": "{MySQL.password}",
+ "database.server.id": "{Debezium.server.id}",
+ "database.server.name": "company_db_{baseID}",
+
+ "database.history.kafka.bootstrap.servers": "{kafka.servers}",
+ "database.history.kafka.topic": "Debezium-history-company-db{baseID}",
+ "database.history.store.only.monitored.tables.ddl": true,
+ "database.history.skip.unparseable.ddl": "true",
+
+ "ddl.parser.mode": "antlr",
+ "snapshot.mode": "schema_only",
+
+ "snapshot.locking.mode": "none",
+ "only.in.env.test.snapshot.locking.mode": "minimal",
+
+ "gtid.new.channel.position": "earliest"
+}
+----
+
+== Connector health monitoring and automatic restart.
+
+As Debezium connector can’t automatically recover from MySQL connection issues, it needs another management script to restart the task. At first we used external cron like script for this, but now as nodejs is already running in container, we delegated this responsibility there as well.
+
+The logic is very simple - every 30 seconds it connects to the Kafka Connect REST API running in the local container and retrieves the connector status. If the connector or task is in FAILED state, it tries to restart it by posting connectors/jobname/tasks/1/restart up to 5 times. If still doesn’t work or there’s UNASSIGNED state, starts graceful shutdown of the Docker container.
+
+== Securing the Kafka Connect REST API
+Because the Kafka Connect REST API had no built in security (in the beginning of 2018 at least), anybody could query and see Kafka cluster and MySQL credentials in plaintext. To prevent such leak we’ve added simple nodejs http proxy script with http basic auth to different port and only expose this port to outside. Once auth is passed this nodejs proxy just forwards all the requests to kafka connect local port 8080.
```

Comment:
```suggestion
Because the Kafka Connect REST API had no built in security (in the beginning of 2018 at least), anybody could query and see Kafka cluster and MySQL credentials in plaintext. To prevent such leak, we’ve added a simple Node.js HTTP proxy script with HTTP basic auth on a different port and only expose this port to the outside. Once auth is passed this Node.js proxy just forwards all the requests to Kafka Connect on local port 8080.
```